core/src/Revolution/modX.php

```diff
@@ -1703,88 +1702,90 @@ public function invokeEvent($eventName, array $params= array ()) {
-        if (isset($options['location']) && !empty($options['location'])) $processorsPath .= ltrim($options['location'],'/') . '/';
 
         // Prevent path traversal through the action
         $action = preg_replace('/[\.]{2,}/', '', htmlspecialchars($action));
 
-        // Find the processor file, preferring class based processors over old-style processors
-        $processorFile = $processorsPath.ltrim($action . '.class.php','/');
+        // Find a matching processor file
+        $processorFile = $processorsPath . ltrim($action . '.class.php','/');
+
         if (!file_exists($processorFile)) {
-            $processorFile = $processorsPath.ltrim($action . '.php','/');
-            $isClass = false;
-        }
-
-        // Prepare a response
-        $response = '';
-        if (file_exists($processorFile)) {
-            if (!isset($this->lexicon)) $this->services->get('lexicon');
-            if (!isset($this->error)) $this->services->get('error');
-
-            if ($isClass) {
-                /* ensure processor file is only included once if run multiple times in a request */
-                if (!array_key_exists($processorFile,$this->processors)) {
-                    $className = include_once $processorFile;
-                    /* handle already included core classes */
-                    if ($className == 1) {
-                        $s = explode('/',$action);
-                        $o = array();
-                        foreach ($s as $k) {
-                            $o[] = ucfirst(str_replace(array('.','_','-'),'',$k));
-                        }
-                        $className = 'mod'.implode('',$o).'Processor';
-                    }
-                    $this->processors[$processorFile] = $className;
-                } else {
-                    $className = $this->processors[$processorFile];
-                }
-                if (!empty($className)) {
-                    $processor = call_user_func_array(array($className,'getInstance'),array(&$this,$className,$scriptProperties));
+            $this->log(modX::LOG_LEVEL_ERROR, "Unable to load processor for action \"{$action}\", it does not exist as an autoloadable class that extends \MODX\Revolution\Processors\Processor, and also not as a file in \"{$processorFile}\"");
+            return '';
+        }
+
+        // Check if we already know the name for the class (meaning we've loaded it before in this request)
+        if (array_key_exists($processorFile,$this->processors)) {
+            $className = $this->processors[$processorFile];
+        }
+        // Load the file. The file is expected to return (as a string) the name of its class.
+        else {
+            $className = include_once $processorFile;
+            // include_once returns 1 if there was no return value in the file, or true if it was already loaded before
+            // In both cases we can guess the class name based on the provided action
+            if ($className == 1) {
```

Comment:
To make it more strict:
```suggestion
            if ($className === 1 || $className === true) {
```